### PR TITLE
[scudo] Make BufferPool typed and always express sizes in terms of element count

### PR DESCRIPTION
**Pull request by fabio-d**
This fixes the issue that resulted in getBuffer interpreting its
argument as a number of elements and getDynamicBuffer interpreting it
as a number of bytes.

### CODE REVIEW

**Comment by hctim** [comment]
Looks reasonable to me, two things:

1. Seems unnecessary to template the `uptr` type of the buffer class, unless you can elaborate on why you need that?
2. Can you please add a small justification in your commit message about how you discovered the mismatch? It's nice to know the backstory here.

**Comment by fabio-d** [comment]
Hi @hctim, thanks for reviewing :) I'll reply in reverse order:

2 -> Unfortunately I don't have a nice backstory as I'm not aware of any nasty issues causes by this. :smile: I've just found the mismatch while migrating BufferPool to the new platform abstraction layer (i.e. from `map`/`unmap` to `MemMap`, see #66788)

1 -> Yes, the only user is `RegionPageMap`, in the same `.h` file, and it only instantiates it with `uptr` elements. The idea to use a template came up while discussing with @ChiaHungDuan on whether to fix the issue by uniformly interpreting arguments as a number of bytes or as a number of elements.
While we initially considered always interpreting them as a number of bytes and returning `void*` from `getBuffer`, we realized that that it would have made the class harder to use (because the caller would be responsible for always requesting a size that is multiple of the intended element size) and also to implement (because the underlying storage would have had to be a simple byte array, to avoid aliasing issues, while suitably aligned for the intended type). The conclusion was that requesting the type as a template and always interpreting sizes as number of elements make both things simpler.

**Comment by hctim** [comment]
> 1 -> Yes, the only user is `RegionPageMap`, in the same `.h` file, and it only instantiates it with `uptr` elements. The idea to use a template came up while discussing with @ChiaHungDuan on whether to fix the issue by uniformly interpreting arguments as a number of bytes or as a number of elements. While we initially considered always interpreting them as a number of bytes and returning `void*` from `getBuffer`, we realized that that it would have made the class harder to use (because the caller would be responsible for always requesting a size that is multiple of the intended element size) and also to implement (because the underlying storage would have had to be a simple byte array, to avoid aliasing issues, while suitably aligned for the intended type). The conclusion was that requesting the type as a template and always interpreting sizes as number of elements make both things simpler.

IMHO there's enough templates already floating around in scudo. If there's no plans to add additional callsites, and it seems fine that `getBuffer` can return `uptr*`, seems like a good idea to just remove the template, and make this patch a pure migration from `BufferSize -> NumElements`.

**Comment by ChiaHungDuan** [comment]
There's plan for different types in BufferPool but I agree that we can stick on uptr at this moment

File: compiler-rt/lib/scudo/standalone/release.h [hunk 1]
```diff
@@ -200,22 +208,16 @@ template <uptr StaticBufferCount, uptr StaticBufferSize> class BufferPool {
...
-        CounterSizeBitsLog(0),
-        CounterMask(0),
-        PackingRatioLog(0),
-        BitOffsetMask(0),
-        SizePerRegion(0),
-        BufferSize(0),
-        Buffer(nullptr) {}
+      : Regions(0), NumCounters(0), CounterSizeBitsLog(0), CounterMask(0),
+        PackingRatioLog(0), BitOffsetMask(0), SizePerRegion(0),
+        BufferNumElements(0), Buffer(nullptr) {}
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/release.h:213]: Just curious, are these lines formatted by clang-format? I think you didn't change the code here.

**Review by ChiaHungDuan — COMMENTED** [review]
Could you also update the title?

**Review by ChiaHungDuan — APPROVED** [review]
Feel free to submit it after the title/commit message updated

File: compiler-rt/lib/scudo/standalone/release.h (hunk 1 above)
Comment (fabio-d) [reply compiler-rt/lib/scudo/standalone/release.h:213]: I did change `BufferSize` to `BufferNumElements`, and the GH presubmit checks complained:
Screenshot from 2023-09-27 15-55-16
The fix was to run clang-format, yes

**Pull request merged**